protocols/kad/src/behaviour.rs

```diff
@@ -2674,6 +2683,12 @@ pub enum KademliaEvent {
     /// See [`Kademlia::kbucket`] for insight into the contents of
     /// the k-bucket of `peer`.
     PendingRoutablePeer { peer: PeerId, address: Multiaddr },
+
+    /// This peer's mode has been updated automatically.
+    ///
+    /// This happens in response to an external
+    /// address being added or removed.
+    ModeChanged { new_mode: Mode },
```

Comment:
Yeah this is a breaking change unfortunately because somebody who matches on this exhaustively will get a compile error now.

Reply:
Our CI even caught that :D

https://github.com/libp2p/rust-libp2p/actions/runs/6244140740/job/16950628734?pr=4503

Reply:
What about marking the KademliaEvent enum as `#[non_exhaustive]` to help mitigate that? According to the RFC for that attribute, inside the crate it is ignored but downstream dependents are required to add a wildcard (`_ => ...`) arm on any `match` on the enum. Most of our example code and my own code already has wildcard arms for processing events from libp2p. I think if we made this switch it would largely be a one-time breaking change and we could avoid having to bump the minor number any time we add an event.

Reply:
> What about marking the KademliaEvent enum as `#[non_exhaustive]` to help mitigate that? According to the RFC for that attribute, inside the crate it is ignored but downstream dependents are required to add a wildcard (`_ => ...`) arm on any `match` on the enum. Most of our example code and my own code already has wildcard arms for processing events from libp2p. I think if we made this switch it would largely be a one-time breaking change and we could avoid having to bump the minor number any time we add an event.

We considered that at some point but I am not sure it is such a good idea on the whole for our usecase. `rust-libp2p` has many moving parts and with its asynchronous operation model with emitted events, grasping the data flow is sometimes difficult. I'd like to allow users to exhaustively match on an enum to allow the compiler to tell them that they are now missing a new variant. There is an unstable lint that users could activate to get notified of a new variant that is now present: https://github.com/rust-lang/rust/issues/89554

I am open to changing our view on this but I think it needs to be carefully considered. Mind opening a discussion about this?

Reply:
Moved discussion to here: #4543